rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+2.  Support for partitioned variable in `ParameterServerStrategy` in eager mode is not a near-term goal since parameter server support in eager runtime won't be ready in the near future.
+3.  We will not support any flavor of online learning or liquid sharding, e.g. dynamically placing or resizing partitions.
+
+
+## Design Proposal
+
+
+### Overview
+
+
+Since Distribution Strategy is a library that is designed to parallelize and scale up models. We think that support for embeddings and partitioned layers should exist in Distribution Strategys. We'll support them differently for different Distribution Strategy:
+
+|          |Embeddings|Partitioned Layer (for Loadbalance Only)|
+|:-------- |:-------- |:----------------|
+|Mirrored/ CollectiveAllReduce Strategy|Pre-TF 2.0: <br>w/ Keras' Embedding layer API: won't be sharded but will be mirrored on each replica. <br><br>Post-TF 2.0: <br>w/ Distribution Strategy's API directly: possibly allow sharding; allow being placed on host; or dense update. <br><br>`Heuristics` in Embedding layer to auto-set performance improvement hints.|Pre-TF 2.0: <br>w/ Keras' layer API: won't be sharded but will be mirrored on all replicas, just like a normal layer. <br><br>Post-TF 2.0: w/ Distribution Strategy's API directly: allow being mirrored on each host. <br><br>Heuristics in Embedding layer to auto-place larger layers on host.|
+|ParameterServer Strategy|Pre-TF 2.0:<br>w/ Keras' Embedding layer API: probably sharded, according to the `partitioner` passed to its constructor.<br><br>Post-TF 2.0:<br>w/ Distribution Strategy's API directly: allow creating embedding variables with different partitioners.|Pre-TF 2.0:<br>w/ Keras' layer API: probably sharded, according to the `partitioner` passed to its constructor which applies globally to all layers.|
+
+In the `ParameterServerStrategy`, the goal of supporting partitioned variable is to match the existing behavior in 1.x. We will propose its APIs for TF 2.0 below. 
+
+**In the mirrored and collective allreduce architecture, we will only mirror embeddings or logits layers, either on devices or on host.** 
+
+In the future, we will probably support a new strategy that is a hybrid of these two architectures in the future.
+
+Please refer to the discussion in [Alternatives Considered](#Alternatives-Considered) section for why we've made these decisions.
+
+
+### Proposed Change for pre-TF 2.0
+
+
+The following is planned to be done before TF 2.0's release.
+
+
+#### A new type `ShardedVariable`
+
+
+As discussed above, `PartitionedVariable` concatenates component variables silently which can only serve the purpose of loadbalancing. So we need another object for uses cases where computation needs to be parellelized or storage needs to be sharded. Users don't have to deal with it as long as they use Keras' layer API and it won't be exposed as a public API in pre-TF 2.0 stage.
+
+```python
+class ShardedVariable(object):
```

Comment:
Will it be better by providing some specific use cases with SharededVariable to demonstrate its difference against the PartitionedVariable?

Reply:
The `ShardedVariable` is just like `ParititionedVariable` excepts it doesn't auto-concats partitions.  It will be passed to embedding lookup and a few loss functions. So it is pretty dummy.